File: cpp_bind/hacd.cpp
#include "hacd.h"
#include "../Extras/HACD/hacdHACD.h"
#include <vector>

nkHACDHandle nk_hacd_create()
{
    HACD::HACD* hacd = new HACD::HACD();
    return static_cast<nkHACDHandle>(hacd);
}

void nk_hacd_destroy(nkHACDHandle hacd)
{
    if (!hacd) return;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    delete hacdInst;
}

void nk_hacd_set_params(nkHACDHandle hacd, const nkHACDParams* params)
{
    if (!hacd || !params) return;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    hacdInst->SetNClusters(params->maxHullCount);
    hacdInst->SetNVerticesPerCH(params->maxVerticesPerHull);
    hacdInst->SetConcavity(params->concavity);
    hacdInst->SetCompacityWeight(params->alpha);
    hacdInst->SetVolumeWeight(params->beta);
    hacdInst->SetConnectDist(params->ccConnectDist);
    hacdInst->SetAddFacesPoints(params->addFacesPoints != 0);
    hacdInst->SetAddExtraDistPoints(params->addExtraDistPoints != 0);
    hacdInst->SetAddNeighboursDistPoints(params->addNeighboursDistPoints != 0);
}

void nk_hacd_get_params(nkHACDHandle hacd, nkHACDParams* outParams)
{
    if (!hacd || !outParams) return;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    outParams->maxHullCount = static_cast<int>(hacdInst->GetNClusters());
    outParams->maxVerticesPerHull = static_cast<int>(hacdInst->GetNVerticesPerCH());
    outParams->concavity = hacdInst->GetConcavity();
    outParams->alpha = hacdInst->GetCompacityWeight();
    outParams->beta = hacdInst->GetVolumeWeight();
    outParams->ccConnectDist = hacdInst->GetConnectDist();
    outParams->addFacesPoints = hacdInst->GetAddFacesPoints() ? 1 : 0;
    outParams->addExtraDistPoints = hacdInst->GetAddExtraDistPoints() ? 1 : 0;
    outParams->addNeighboursDistPoints = hacdInst->GetAddNeighboursDistPoints() ? 1 : 0;
}

int nk_hacd_set_points(nkHACDHandle hacd, const double* points, int numPoints)
{
    if (!hacd || !points || numPoints <= 0) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    std::vector<HACD::Vec3<HACD::Real>>* vecPoints = new std::vector<HACD::Vec3<HACD::Real>>(numPoints);
    for (int i = 0; i < numPoints; i++)
    {
        (*vecPoints)[i] = HACD::Vec3<HACD::Real>(points[i * 3], points[i * 3 + 1], points[i * 3 + 2]);
    }
    
    hacdInst->SetPoints(&(*vecPoints)[0]);
    hacdInst->SetNPoints(numPoints);
    
    return 1;
}

int nk_hacd_set_triangles(nkHACDHandle hacd, const int* triangles, int numTriangles)
{
    if (!hacd || !triangles || numTriangles <= 0) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    std::vector<HACD::Vec3<long>>* vecTriangles = new std::vector<HACD::Vec3<long>>(numTriangles);
    for (int i = 0; i < numTriangles; i++)
    {
        (*vecTriangles)[i] = HACD::Vec3<long>(triangles[i * 3], triangles[i * 3 + 1], triangles[i * 3 + 2]);
    }
    
    hacdInst->SetTriangles(&(*vecTriangles)[0]);
    hacdInst->SetNTriangles(numTriangles);
    
    return 1;
}

int nk_hacd_compute(nkHACDHandle hacd)
{
    if (!hacd) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    hacdInst->NormalizeData();
    bool result = hacdInst->Compute(false, false);
    
    return result ? 1 : 0;
}

int nk_hacd_get_num_hulls(nkHACDHandle hacd)
{
    if (!hacd) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    return static_cast<int>(hacdInst->GetNClusters());
}

int nk_hacd_get_hull_points_count(nkHACDHandle hacd, int hullIndex)
{
    if (!hacd) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    return static_cast<int>(hacdInst->GetNPointsCH(hullIndex));
}

int nk_hacd_get_hull_triangles_count(nkHACDHandle hacd, int hullIndex)
{
    if (!hacd) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    return static_cast<int>(hacdInst->GetNTrianglesCH(hullIndex));
}

int nk_hacd_get_hull(nkHACDHandle hacd, int hullIndex, double* outPoints, int* outNumPoints, int* outTriangles, int* outNumTriangles)
{
    if (!hacd || hullIndex < 0) return 0;
    HACD::HACD* hacdInst = static_cast<HACD::HACD*>(hacd);
    
    size_t numPoints = hacdInst->GetNPointsCH(hullIndex);
    size_t numTriangles = hacdInst->GetNTrianglesCH(hullIndex);
    
    if (numPoints == 0 || numTriangles == 0) return 0;
    
    std::vector<HACD::Vec3<HACD::Real>> points(numPoints);
    std::vector<HACD::Vec3<long>> triangles(numTriangles);
    
    if (!hacdInst->GetCH(hullIndex, &points[0], &triangles[0])) return 0;
    
    if (outPoints && outNumPoints)
    {
        for (size_t i = 0; i < numPoints; i++)
        {
            outPoints[i * 3] = points[i].X();
            outPoints[i * 3 + 1] = points[i].Y();
            outPoints[i * 3 + 2] = points[i].Z();
        }
        *outNumPoints = static_cast<int>(numPoints);
    }
    
    if (outTriangles && outNumTriangles)
    {
        for (size_t i = 0; i < numTriangles; i++)
        {
            outTriangles[i * 3] = static_cast<int>(triangles[i].X());
            outTriangles[i * 3 + 1] = static_cast<int>(triangles[i].Y());
            outTriangles[i * 3 + 2] = static_cast<int>(triangles[i].Z());
        }
        *outNumTriangles = static_cast<int>(numTriangles);
    }
    
    return 1;
}

File: rust/src/core/hacd.rs
use std::ffi::c_void;
use std::ptr::NonNull;

use crate::ffi;

pub struct HACDParams {
    pub max_hull_count: i32,
    pub max_vertices_per_hull: i32,
    pub concavity: f64,
    pub alpha: f64,
    pub beta: f64,
    pub cc_connect_dist: f64,
    pub add_faces_points: bool,
    pub add_extra_dist_points: bool,
    pub add_neighbours_dist_points: bool,
}

impl Default for HACDParams {
    fn default() -> Self {
        Self {
            max_hull_count: 32,
            max_vertices_per_hull: 64,
            concavity: 0.0025,
            alpha: 0.05,
            beta: 0.05,
            cc_connect_dist: 30.0,
            add_faces_points: true,
            add_extra_dist_points: false,
            add_neighbours_dist_points: false,
        }
    }
}

impl HACDParams {
    fn to_ffi(&self) -> ffi::nkHACDParams {
        ffi::nkHACDParams {
            maxHullCount: self.max_hull_count,
            maxVerticesPerHull: self.max_vertices_per_hull,
            concavity: self.concavity,
            alpha: self.alpha,
            beta: self.beta,
            ccConnectDist: self.cc_connect_dist,
            addFacesPoints: if self.add_faces_points { 1 } else { 0 },
            addExtraDistPoints: if self.add_extra_dist_points { 1 } else { 0 },
            addNeighboursDistPoints: if self.add_neighbours_dist_points { 1 } else { 0 },
        }
    }
}

pub struct ConvexHull {
    pub points: Vec<f64>,
    pub triangles: Vec<i32>,
}

impl ConvexHull {
    pub fn num_points(&self) -> i32 {
        (self.points.len() / 3) as i32
    }

    pub fn num_triangles(&self) -> i32 {
        (self.triangles.len() / 3) as i32
    }

    pub fn get_point(&self, index: i32) -> [f64; 3] {
        let base = (index * 3) as usize;
        [self.points[base], self.points[base + 1], self.points[base + 2]]
    }

    pub fn get_triangle(&self, index: i32) -> [i32; 3] {
        let base = (index * 3) as usize;
        [self.triangles[base], self.triangles[base + 1], self.triangles[base + 2]]
    }
}

pub struct HACD {
    handle: NonNull<c_void>,
}

impl HACD {
    pub fn new() -> Self {
        let handle = unsafe { ffi::nk_hacd_create() };
        Self {
            handle: NonNull::new(handle).expect("Failed to create HACD instance"),
        }
    }

    pub fn set_params(&self, params: &HACDParams) {
        let ffi_params = params.to_ffi();
        unsafe {
            ffi::nk_hacd_set_params(self.handle.as_ptr(), &ffi_params);
        }
    }

    pub fn get_params(&self) -> HACDParams {
        let mut ffi_params: ffi::nkHACDParams = unsafe { std::mem::zeroed() };
        unsafe {
            ffi::nk_hacd_get_params(self.handle.as_ptr(), &mut ffi_params);
        }
        HACDParams {
            max_hull_count: ffi_params.maxHullCount,
            max_vertices_per_hull: ffi_params.maxVerticesPerHull,
            concavity: ffi_params.concavity,
            alpha: ffi_params.alpha,
            beta: ffi_params.beta,
            cc_connect_dist: ffi_params.ccConnectDist,
            add_faces_points: ffi_params.addFacesPoints != 0,
            add_extra_dist_points: ffi_params.addExtraDistPoints != 0,
            add_neighbours_dist_points: ffi_params.addNeighboursDistPoints != 0,
        }
    }

    pub fn set_mesh(&self, points: &[f64], triangles: &[i32]) -> Result<(), String> {
        let num_points = (points.len() / 3) as i32;
        let num_triangles = (triangles.len() / 3) as i32;

        let points_result = unsafe {
            ffi::nk_hacd_set_points(self.handle.as_ptr(), points.as_ptr(), num_points)
        };

        if points_result != 0 {
            return Err("Failed to set points".to_string());
        }

        let triangles_result = unsafe {
            ffi::nk_hacd_set_triangles(self.handle.as_ptr(), triangles.as_ptr(), num_triangles)
        };

        if triangles_result != 0 {
            return Err("Failed to set triangles".to_string());
        }

        Ok(())
    }

    pub fn compute(&self) -> Result<(), String> {
        let result = unsafe { ffi::nk_hacd_compute(self.handle.as_ptr()) };
        if result != 0 {
            Err("HACD computation failed".to_string())
        } else {
            Ok(())
        }
    }

    pub fn get_num_hulls(&self) -> i32 {
        unsafe { ffi::nk_hacd_get_num_hulls(self.handle.as_ptr()) }
    }

    pub fn get_hull(&self, hull_index: i32) -> Option<ConvexHull> {
        let num_points = unsafe {
            ffi::nk_hacd_get_hull_points_count(self.handle.as_ptr(), hull_index)
        };
        let num_triangles = unsafe {
            ffi::nk_hacd_get_hull_triangles_count(self.handle.as_ptr(), hull_index)
        };

        if num_points <= 0 || num_triangles <= 0 {
            return None;
        }

        let mut points = vec![0.0f64; (num_points * 3) as usize];
        let mut triangles = vec![0i32; (num_triangles * 3) as usize];
        let mut out_num_points: i32 = 0;
        let mut out_num_triangles: i32 = 0;

        let result = unsafe {
            ffi::nk_hacd_get_hull(
                self.handle.as_ptr(),
                hull_index,
                points.as_mut_ptr(),
                &mut out_num_points,
                triangles.as_mut_ptr(),
                &mut out_num_triangles,
            )
        };

        if result != 0 {
            return None;
        }

        points.truncate((out_num_points * 3) as usize);
        triangles.truncate((out_num_triangles * 3) as usize);

        Some(ConvexHull { points, triangles })
    }

    pub fn get_hulls(&self) -> Vec<ConvexHull> {
        let num_hulls = self.get_num_hulls();
        let mut hulls = Vec::with_capacity(num_hulls as usize);

        for i in 0..num_hulls {
            if let Some(hull) = self.get_hull(i) {
                hulls.push(hull);
            }
        }

        hulls
    }
}

impl Default for HACD {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for HACD {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_hacd_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for HACD {}
unsafe impl Sync for HACD {}
